adr/adr.001.folder structure/adr.001.folder structure.md

```diff
@@ -0,0 +1,260 @@
+# ADR-001: Feature-Based Folder Structure
+
+## Status
+Proposed
+
+## Date
+- **Proposed**: 2024-10-16
+- **Decided**: 
+- **Work Began**: 
+- **Work Completed**: 
+
+## Context
+
+Our current codebase is organized mainly by the type of files and the framework components, such as `components`, `data`, `hooks`, `screens`, `utils`, etc. While this structure groups similar types of files together, it scatters related functionality across multiple directories. This can make navigation and maintenance challenging, especially as the codebase grows.
+
+We are increasingly relying on AI-assisted tooling for development. Having code that changes together located in different parts of the codebase reduces the effectiveness of such tools. Packaging by feature would enhance productivity by allowing developers (and AI tools) to focus on a single directory containing all relevant code for a feature.
```

Comment:
This is a great point. I wonder what other evolutions in coding practice we'll see to optimize for AI assistance.

Reply:
Architecture! Architecture! Architecture! 

We're absolutely going to see a return to sound software engineering fundamentals as the essence of software has not changed for approximately 60 years but we've just got distracted by all the shiny new toys we've lost track of how to build it in a maintainable fashion.

Reply:
Right now the problem is all LLM is trained on 90% bad code.

Something that you could do @technoplato that would be AMAZING is create some .md files somewhere where we can tag those files for reference.

So in Cursor I would just do refactor like this file @THE_FILE

We can start with just one file that lay out all the best practices like this https://x.com/guidsen/status/1830289361801470118

And then I would take a look at this https://x.com/zbeyens/status/1832079140083687671

Reply:
Oh this is great. Will have a look.

Reply:
> Right now the problem is all LLM is trained on 90% bad code.

1,000% agree. With tools like Cursor like you mention though, with solid fundamentals, we can "override" those poor tendencies to follow our ideal standards with a couple examples. We're just going to have to coach the thing through designing the feature based on our standards. 

> Something that you could do @technoplato that would be AMAZING is create some .md files somewhere where we can tag those files for reference.

This is an awesome idea. 

https://docs.cursor.com/context/rules-for-ai

<details><summary>copying that tweets example for when the tweet link breaks: 
</summary>
<p>

You are an expert in TypeScript, Node.js, Next.js App Router, React, Shaden UI,
Radix UI, Supabase, and Tailwind.
Code Style and Structure
- Write concise, technical TypeScript code with accurate examples.
- Use functional and declarative programming patterns; avoid classes.
- Prefer iteration and modularization over code duplication.
- Use descriptive variable names with auxiliary verbs (e.g., isLoading, hasError). - Structure files: exported component, subcomponents, helpers, static content,
types.
Naming Conventions
- Use lowercase with dashes for directories (e.g., components/auth-wizard).
- Favor named exports for components.
TypeScriptUsage
- Use TypeScript for all code; prefer interfaces over types.
- Avoid enums; use maps instead.
- Use functional components with TypeScript interfaces.
Syntax and Formatting
- Use the "function" keyword for pure functions. - Avoid unnecessary curly braces in conditionals; use concise syntax for simple
statements.
- Use declarative JSX.
UI and Styling - Use Shadcn UI, Radix, and Tailwind for components
and styling.
- Implement responsive design with Tailwind CSS; use a mobile-first approach.
Performance Optimization - Minimize 'use client', 'useEffect', and 'setState'; favor React Server Components
(RSC) •
- Wrap client components in Suspense with fallback.
- Use dynamic loading for non-critical components.
- Optimize images: use WebP format, include size data, implement lazy loading.
Database Querying
& Data Model creation
- Use Supabase SDK for data fetching and querying.
- For data model creation, use Supabase's schema builder.
Key Conventions
- Use 'nuqs' for URL search parameter state management.
- Optimize Web Vitals (LCP, CLS, FID).
- Limit 'use client':
- Favor server components and Next.js SSR.
- Use only for Web API access in small components.
- Avoid for data fetching or state management.
Follow Next.js docs for Data
Fetching, Rendering, and Routing.

</p>
</details> 


> We can start with just one file that lay out all the best practices like this https://x.com/guidsen/status/1830289361801470118

This looks fairly straightforward and will get us some momentum, especially with new architecture once it is in place. 

> And then I would take a look at this https://x.com/zbeyens/status/1832079140083687671

This is a very cool idea and I think we can definitely use some pieces from it!